src/Draco.Compiler/Syntax/ParseTree.cs

```diff
@@ -0,0 +1,144 @@
+using Draco.Compiler.Utilities;
+
+namespace Draco.Compiler.Syntax;
+
+/// <summary>
+/// An immutable structure representing a parsed source text with information about concrete syntax.
+/// </summary>
+internal abstract record class ParseTree
+{
+    public sealed record class CompilationUnit(
+        ValueArray<Decl> Declarations) : ParseTree;
+
+    public abstract record class Decl : ParseTree
+    {
+        public sealed record class Func(
+            IToken FuncKeyword,
+            IToken Identifier,
+            IToken OpenParenToken,
+            ValueArray<(
+                (   IToken Identifier,
+                    TypeSpecifier Type
+                ) Param,
+                IToken CommaToken
+            )> Params,
+            IToken CloseParenToken,
+            TypeSpecifier? Type,
+            FuncBody Body) : Decl;
+
+        public sealed record class Label(
+            IToken Identifier,
+            IToken ColonToken) : Decl;
+
+        public sealed record class Variable(
+            IToken Keyword, // Either var or val
+            IToken Identifier,
+            TypeSpecifier? Type,
+            (   IToken EqualsToken,
+                Expr Expression
+            )? Initializer) : Decl;
```

Comment:
I'm generally not a fan of tuples in datastructures, but I also can't think of anything more sensible other than defining this construct separately, which would be annoying and useless. It can stay for now.

Reply:
Was kind of the only good/concise way I could think of having several values be nullable together but not separately.